js/utils/pseudocolumn_helpers.js

```diff
@@ -215,116 +196,200 @@
-                res, i, term, col, path, ds, constraints, parsed, useRightJoin;
+        },
 
-            // go through list of facets
-            for (i = 0; i < and.length; i++) {
-                term = and[i];
+        getErrorOutput: function (message, index) {
+           return {successful: false, message: message + "(index=" + index +")"};
+       }
 
-                if (typeof term !== "object") {
-                    return getErrorOutput(facetErrors.invalidFacet, i);
-                }
+    };
 
-                // get the column name
-                col = _getSourceColumnStr(term.source);
-                if (typeof col !== "string") {
-                    return getErrorOutput(facetErrors.invalidSource, i);
-                }
+    /**
+     * Given a facet and extra needed attributes, return the equivalent ermrest query
+     * For the structure of JSON, take a look at ParsedFacets documentation.
+     *
+     * NOTE:
+     * - If any part of the facet is not as expected, it will throw an object
+     *   with ``"successful": false` and an appropriate `message`.
+     *   Any part of the code that is using this function should guard against the result
+     *   being empty, and throw error in that case.
+     *
+     * - We are not handling multiple `null` in the facet and it will restult in an error.
+     *
+     * - If we encounter a `null` filter, we are going to change the structure of url.
+     *   Assume that this facet is for table `A` and table `B` has null filter. Therefore
+     *   the url will be:
+     *      S:B/<filters of B>/<path from B to A where the last join is right join>/<parsed filter of all the other facets>/$alias
+     *   Compare this with the following which will be the result if none of the facets have `null`:
+     *      <parsed fitler of all the facets>/$alias
+     *   In this case, we are returning a `"rightJoin": true`. In this case, we
+     *   have to make sure that the returned value must be the start of url and
+     *   cannot be simply appended to the rest of url.
+     *
+     * @param       {object} json  JSON representation of filters
+     * @param       {string} alias the table alias
+     * @param       {string} schemaName the starting schema name
+     * @param       {string} tableName the starting table name
+     * @param       {string} catalogId the catalog id
+     * @param       {ERMrest.catalog} [catalogObject] the catalog object (could be undefined)
+     * @param       {Object[]} [consNames] the constraint names (could be undefined)
+     * @constructor
+     * @return      {object} An object that will have the following attributes:
+     * - successful: Boolean (true means successful, false means cannot be parsed).
+     * - parsed: String  (if the given string was parsable).
+     * - message: String (if the given string was not parsable)
+     */
+    _renderFacet = function(json, alias, schemaName, tableName, catalogId, catalogObject, consNames) {
```

Comment:
The diff for this file is a bit goofy. It looks like this is where you changed `_JSONToErmrestFilter` to `_renderFact`. It looks like you also moved functions that were scoped within that function to be outside of it for reuse. I don't see many changes to the actual logic though besides explicitly checking for the `SEARCH_BOX` source key.

Reply:
That's right. The only actual change to this function is supporting `sourcekey` which includes the `SEARCH_BOX` support.